Filterable select

atmgrifter00 opened the pull request:
# Pull Request

## 🤨 Rationale

- #1273
- [Filterable Select HLD](https://github.com/ni/nimble/blob/main/packages/nimble-components/src/select/specs/filterable-select-hld.md)

## 👩‍💻 Implementation

As the behaviors of the filtering resemble the `Combobox` filtering, I have brought over the necessary implementation from the FAST `Combobox` to deal with filtered options (albeit slightly modified as necessary). Since the `Select` will now potentially display new user-visible text, I have added two new core label provider APIs: `selectFilterSearch` and `selectFilterNoResults`.

Some important new behaviors I am also introducing are:
- The display value will no longer update as a user navigates between options in the dropdown
- Pressing `Esc` will no longer commit a value

Both of these changes are aligned with [ARIA best practices](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-select-only/#accessibilityfeatures), despite them not being aligned with the native select behavior.

We have also now essentially forked FAST's `Select` implementation and no longer derive from it, but our own version of [their `FormAssociatedSelect`](https://github.com/microsoft/fast/blob/c64110d2f1dfbc24a0978c362bcf72b8958e6d73/packages/web-components/fast-foundation/src/select/select.form-associated.ts#L12). This was thought prudent with the large number of modifications we had to put in place in various method overrides, and with an eye towards even more changes in the future.

## 🧪 Testing

New unit tests. New matrix tests.

## ✅ Checklist

<!--- Review the list and put an x in the boxes that apply or ~~strike through~~ around items that don't (along with an explanation). -->

- [ ] I have updated the project documentation to reflect my changes or determined no changes are needed.

mollykreis commented on packages/nimble-components/src/select/tests/doc-strings.ts:
```diff
@@ -0,0 +1 @@
+export const filterModeDescription = "This attribute controls the filtering behavior of the `Select`. The default is `none` results in a dropdown with no input for filtering. A non-'none' setting results in a search input placed at the top or the bottom of the dropdown when openend (depending on where the popup is shown relative to the component). The `standard` setting will perform a case-insens ...
```
```suggestion
export const filterModeDescription = "This attribute controls the filtering behavior of the `Select`. The default of `none` results in a dropdown with no input for filtering. A non-'none' setting results in a search input placed at the top or the bottom of the dropdown when opened (depending on where the popup is shown relative to the component). The `standard` setting will perform a case-insensitive and diacritic-insensitive filtering of the available options anywhere within the text of each option.";
```

mollykreis commented on packages/nimble-components/src/select/styles.ts:
```diff
@@ -28,6 +40,93 @@ export const styles = css`
         display: contents;
     }
+
+    .listbox {
```
Can you take a look at the Color theme styling? It doesn't look right -- the semi-transparent background color no longer goes all the way to the edges of the dropdown.

Published storybook:
![image](https://github.com/ni/nimble/assets/20542556/36b53a6b-ee38-4b07-b4c1-05cfbe721138)

This PR's storybook:
![image](https://github.com/ni/nimble/assets/20542556/e30c8346-0a27-43e1-846c-cdb3cec9ab78)

atmgrifter00 replied:
Fixed.

mollykreis commented on packages/nimble-components/src/select/tests/select-opened-matrix.stories.ts:
```diff
@@ -17,16 +18,19 @@ export default metadata;
-    <${selectTag} open position="${() => position}" style="${() => positionStyle}">
+const component = (
+    [position, positionStyle]: PositionState,
+    filterMode: FilterModeState
```
Should there also be stories for the following?
- a filter string being set to make sure the `input` font/color is correct when something other than the placeholder is shown
- the "No items found" state

atmgrifter00 replied:
I've added matrix tests that adds a filter, of which there is no match, and thus shows the 'No items found' text.

mollykreis commented on packages/nimble-components/src/select/tests/select.spec.ts:
```diff
@@ -258,4 +273,142 @@ describe('Select', () => {
+            await pageObject.closeDropdown();
+            await pageObject.clickSelect();
+
+            expect(pageObject.getFilteredOptions().length).toBe(5);
```
Can this test also check that the filter input is blank, not just that the options aren't filtered?

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -67,29 +202,162 @@ export class Select extends FoundationSelect implements ErrorPattern {
+            case ' ': {
+                // when dropdown is open allow user to enter a space for filter text
+                // (calling super method will close dropdown)
+                if (!this.open) {
```
It does...that's the first check in the handler (check if filterMode is `none`, and if it is just return the `super` call).

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -67,29 +202,162 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        }
+
+        if (this.open) {
+            DOM.queueUpdate(() => {
+                window.requestAnimationFrame(() => {
```
While I didn't adopt if from FAST, I did find this comment in the FAST `Listbox` code that suggests I can remove the `DOM.queueUpdate` call in favor of `requestAnimationFrame`, and when I do this it continues to work as expected.

I don't know of another way outside of using a `MutationObserver`, which seems challenging. The `loaded` event for the anchored region does not fire every time you open the dropdown.

atmgrifter00 replied:
The `DOM.queueUpdate` call was removed as it did not seem to be required. At the moment, I haven't found a better means of setting focus on the input when the dropdown is opened.

atmgrifter00 commented on packages/nimble-components/src/select/styles.ts:
```diff
@@ -28,6 +40,93 @@ export const styles = css`
+        width: calc(100% - ${mediumPadding} - (2 * ${borderWidth}));
+    }
+
+    .filter-input {
```
Do you know what this should be? If I style it using `${placeholderFontColor}` it seems to get even dimmer.

mollykreis replied:
Yes, I would expect it to be `${placeholderFontColor}`.

I pulled down your branch and this seems to be working fine for me:
```css
.filter-input::placeholder {
    color: ${placeholderFontColor};
}
```

atmgrifter00 commented on packages/nimble-components/src/select/styles.ts:
```diff
@@ -28,6 +40,93 @@ export const styles = css`
+        padding-right: ${smallPadding};
+    }
+
+    .listbox.empty slot {
+        display: none;
+    }
```
Nope.

atmgrifter00 commented on packages/nimble-components/src/select/styles.ts:
```diff
@@ -28,6 +40,93 @@ export const styles = css`
+
+    .listbox {
+        overflow-x: clip;
+        padding: ${smallPadding};
```
This got removed.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -102,6 +370,28 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        // resolves against all options, since selectedIndex should be relative to entire set.
+        this.filter = '';
+        this.filterOptions();
+        if (selectedItem) {
```
Not needed.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -67,29 +202,162 @@ export class Select extends FoundationSelect implements ErrorPattern {
+                window.requestAnimationFrame(() => {
+                    this.input?.focus();
+                    this.scrollbarIsVisible = this.scrollableElement.clientHeight
+                        < this.scrollableElement.scrollHeight;
```
No longer need this.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -67,29 +202,162 @@ export class Select extends FoundationSelect implements ErrorPattern {
+                break;
+            }
+            case 'Escape': {
+                this.filteredOptions = [];
```
I don't.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,6 +75,115 @@ export class Select extends FoundationSelect implements ErrorPattern {
+    public override get value(): string {
+        Observable.track(this, 'value');
+        // eslint-disable-next-line @typescript-eslint/dot-notation
+        return this['_value'] as string;
```
Done.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,6 +75,115 @@ export class Select extends FoundationSelect implements ErrorPattern {
+     *
+     * @public
+     */
+    public override get value(): string {
```
The getter was removed. Added a comment to the setter.

atmgrifter00 commented on packages/nimble-components/src/select/template.ts:
```diff
@@ -82,20 +88,43 @@ SelectOptions
+                            @input="${(x, c) => x.inputHandler(c.event as InputEvent)}"
+                            @click="${(x, c) => x.inputClickHandler(c.event as MouseEvent)}"
+                            ${ref('input')}
+                            placeholder="${x => selectFilterSearchLabel.getValueFor(x)}..."
```
At the moment I've settled on just "Search" (and thus removed '...' outright).

atmgrifter00 commented on packages/nimble-components/src/select/template.ts:
```diff
@@ -82,20 +88,43 @@ SelectOptions
+                    <div class="search-field ${x => x.positionAttribute}">
+                        <${iconMagnifyingGlassTag}></${iconMagnifyingGlassTag}>
+                        <input
+                            class="filter-input ${x => (x.filter.length === 0 ? 'empty' : '')}"
```
Removed.

atmgrifter00 commented on packages/nimble-components/src/select/testing/select.pageobject.ts:
```diff
@@ -0,0 +1,128 @@
+    public async openAndSetFilterText(filterText: string): Promise<void> {
+        await this.clickSelect();
+        const filterInput = this.getFilterInput();
+        if (filterInput) {
```
Throwing exception now.

mollykreis commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -67,27 +201,164 @@ export class Select extends FoundationSelect implements ErrorPattern {
-    private controlChanged(
-        _prev: HTMLElement | undefined,
-        _next: HTMLElement | undefined
+    // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
```
Can you add a short explanation of why you needed to disable this line rule? Something about how we've set the return type here to match the return type of the super class's keydownHandler.

mollykreis commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -67,27 +201,164 @@ export class Select extends FoundationSelect implements ErrorPattern {
+            this.filteredOptions = this._options;
         }
+
+        this._options.forEach(o => {
+            o.hidden = !this.filteredOptions.includes(o);
+        });
```
A client used to be able to hide an option in the select by marking it as `hidden`. I don't think that works anymore (or at least I couldn't get it to work by modifying the DOM in the storybook page published with this branch).

What are the expectations about this behavior?

atmgrifter00 replied:
Hmmm...good question. The current behavior ends up matching what happens with the `Combobox`, which is to say if a client has marked an option as `hidden` and then proceeds to filter the options, the hidden option will be revealed (and stay revealed even if the filter is removed). It will be good to hear from others on whether this is acceptable.

atmgrifter00 replied:
We discussed this off-line, and we've decided that we will follow the behavior of the `Combobox` but still document it.

mollykreis commented on packages/nimble-components/src/select/styles.ts:
```diff
@@ -28,6 +40,81 @@ export const styles = css`
+        content: '';
+        position: absolute;
+        height: 0px;
+        border-bottom: rgba(${borderRgbPartialColor}, 0.15) 2px solid;
```
Want to double check -- should this be out existing `dividerColor` token?

atmgrifter00 replied:
No. To be clear, there is a `dividerBackgroundColor`, but not `dividerColor`. Additionally, this pattern is used elsewhere for similar visuals like the divider for the Combobox (which is what I copied directly). I'll note that in the designs it actually is spec'd to use 10% and not 15% transparency, so I've updated that part of the style.

mollykreis commented on packages/nimble-components/src/select/styles.ts:
```diff
@@ -28,6 +40,81 @@ export const styles = css`
+        bottom: calc(${controlHeight} + ${smallPadding} + (2 * ${borderWidth}));
+    }
+
+    .search-field:not(.above)::after,
+    .search-field.above::before {
+        width: calc(100% - (2 * ${borderWidth}));
+    }
```
A few questions here:
1. Why can't this `width` styling go above in the larger block of shared `.search-field::after, .search-field::before` styling?
2. It looks like this is attempting to make the gray divider line not span the entire width of the drop-down. Is that correct? It looks that way in the visual designs, but I don't see that being the case in storybook.

mollykreis commented on packages/nimble-components/src/select/styles.ts:
```diff
@@ -28,6 +40,81 @@ export const styles = css`
+        width: 100%;
+    }
+
+    filter-input::placeholder {
```
I can't tell that this styling is working as expected. Shouldn't it be the `filter-input` class that is styled?
```suggestion
    .filter-input::placeholder {
```

mollykreis commented on packages/nimble-components/src/select/template.ts:
```diff
@@ -81,21 +87,49 @@ SelectOptions
+                        ></slot>
+                    </div>
+                    ${when(x => (x.filterMode !== FilterMode.none && x.filteredOptions.length === 0), html<Select>`
+                        <span class="no-results-label ${x => x.positionAttribute}">
```
The alignment of this text seems off to me. It appears to be aligned with the left edge of a list-option rather than with the left edge of the content within a list-option.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,6 +69,120 @@ export class Select extends FoundationSelect implements ErrorPattern {
+
+    // See comment for 'set value' below. Despite the implementation for 'get value'
+    // being an exact copy of the FAST Select one, without replicating it here, we
+    // would see test errors.
```
Basically tests that would access the current `value` to check to see if it was the expected value would fail. At the moment I'm not certain of the reason why having the local implementation is necessary.

mollykreis replied:
I think it would be good to understand why. I wonder if it has to do with the `this` value that ends up being passed to `Observable.track(this, 'value');`.

atmgrifter00 replied:
It was discovered that per the spec, if you provide a setter implementation, you must also provide a getter to avoid a return value of `undefined` (see [this SO post](https://stackoverflow.com/questions/28950760/override-a-setter-and-the-getter-must-also-be-overridden)).

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -67,27 +201,164 @@ export class Select extends FoundationSelect implements ErrorPattern {
     ): void {
         const value = this.value;
         super.slottedOptionsChanged(prev, next);
+        this.filterOptions();
```
I've actually just removed this line. I'm not sure why I added it in the first place (except maybe I thought I was covering some corner-case?). Regardless, I don't think I need this, and certainly no tests fail with it gone.

mollykreis commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -60,34 +198,174 @@ export class Select extends FoundationSelect implements ErrorPattern {
-            this.region.anchorElement = this.control;
+    public override clickHandler(e: MouseEvent): boolean {
+        this.updateSelectedIndexFromFilteredSet();
+        super.clickHandler(e);
+        return true;
```
Would it make sense to return `super.clickHandler` here instead of always returning `true`?

mollykreis commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -60,34 +198,174 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        }
+        this.filterOptions();
+        this.clearSelection();
+        if (this.committedSelectedOption) {
+            this.committedSelectedOption.selected = true;
+        }
```
If the condition of the if-statement directly below (line 245) evaluates to `true`, then the assignment here is undone. Would it make sense to write the logic so that only one executes? Something like:

```ts
if (
    this.filteredOptions.length > 0
    && this.committedSelectedOption
    && !this.filteredOptions.includes(this.committedSelectedOption)
) {
    this.committedSelectedOption.selected = false;
    this.filteredOptions[0]!.selected = true;
} else if (this.committedSelectedOption) {
    this.committedSelectedOption.selected = true;
}
```

mollykreis commented on packages/nimble-components/src/select/template.ts:
```diff
@@ -81,21 +87,49 @@ SelectOptions
+                    ${ref('listbox')}
+                >
+                    ${when(x => x.filterMode !== FilterMode.none, html<Select>`
+                        <div class="search-field ${x => x.positionAttribute}">
```
I think it would be good to standardize on the term "filter", so this would be `class="filter-field"`, and the icon would be `class="filter-icon"`. Thoughts?

mollykreis commented on packages/nimble-components/src/select/testing/select.pageobject.ts:
```diff
@@ -0,0 +1,137 @@
+import { FilterMode } from '../types';
+
+/**
+ * Page object for the `nimble-table-cell` component to provide consistent ways
```
```suggestion
 * Page object for the `nimble-select` component to provide consistent ways
```

mollykreis commented on packages/nimble-components/src/select/tests/select-opened-matrix.stories.ts:
```diff
@@ -45,29 +49,109 @@ if (remaining.length > 0) {
-export const selectAboveOpenDarkThemeBlackBackground: StoryFn = createFixedThemeStory(
-    component(positionStates[1]),
+
+const badFilterPlayFunction = (): void => {
```
nit: This isn't really a "bad" filter (there isn't anything invalid or incorrect about it). What you called it something like "noMatchesFilterPlayFunction" instead?

mollykreis commented on packages/nimble-components/src/select/tests/select.spec.ts:
```diff
@@ -411,5 +414,32 @@ describe('Select', () => {
+            expect(pageObject.getFilteredOptions().length).toBe(3);
+        });
+
+        it('filtering to only disbled item, then pressing <Enter> reverts selection', async () => {
```
```suggestion
        it('filtering to only disabled item, then pressing <Enter> reverts selection', async () => {
```

mollykreis commented on packages/nimble-components/src/select/tests/select.spec.ts:
```diff
@@ -411,5 +414,32 @@ describe('Select', () => {
+            expect(currentSelection?.value).toBe('one');
+        });
+
+        it('filtering to only disbled item, then clicking away reverts selection', async () => {
```
```suggestion
        it('filtering to only disabled item, then clicking away reverts selection', async () => {
```

mollykreis commented on packages/nimble-components/src/select/tests/select.spec.ts:
```diff
@@ -411,5 +414,32 @@ describe('Select', () => {
             expect(pageObject.isNoResultsLabelVisible()).toBeFalse();
         });
+
+        it('clicking disabled option does not cause filtered options to reappear', async () => {
```
How oddly specific this test is makes me laugh since I saw the behavior before you fixed this bug 😄 

However, I'm wondering if it should be something broader like "clicking disabled option does not cause the select to change state". Then, it could check:
- the options stay the same
- the entered filter text stays the same
- the select stays open

mollykreis commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -60,36 +196,205 @@ export class Select extends FoundationSelect implements ErrorPattern {
+                this.focus();
+                break;
+            }
+            case keyEscape: {
```
I haven't reviewed the most recent updates to this PR, but something still seems off with the Escape behavior.

Specifically, I see this:
1. Open the storybook page for the select with filterMode = standard
2. Type a filter that matches something (e.g. "Zur")
3. See that the matching option is highlighted in the dropdown
4. Press Escape
5. Notice that the dropdown closes and reverts back to the original value
6. Click on the select to open it again
7. Notice that the current value is not highlighted in the dropdown and that filtering doesn't cause anything to become highlighted

atmgrifter00 replied:
Yep. Missed a needed subtle change. I've updated and added a test.

jattasNI commented on change/@ni-nimble-components-bcea55bf-ed59-4b11-b80b-6311749da460.json:
```diff
@@ -0,0 +1,7 @@
+{
```
I'm seeing a visual glitch when the drop down position is set to "above". The items that are scrolling behind the filter input are overlapping the line next to the input.

![Screenshot 2024-01-30 at 7 59 37 AM](https://github.com/ni/nimble/assets/10500124/1f59fb25-831a-412b-8f19-30577f4477c6)

jattasNI replied:
Definitely much better. It still looks a bit odd when zoomed way in because the top and bottom of the border behave differently. The top is either further back in the z-order or more transparent, so the items are still visible through/over it, while the bottom causes items to clip. But it's not really noticeable unless zoomed in so I'd say it's acceptable if it's hard to fix.


https://github.com/ni/nimble/assets/10500124/4ee21cec-68a4-46e9-aeb9-006a80e3b978

atmgrifter00 replied:
Made one more style change that I think addresses this.

jattasNI commented on change/@ni-nimble-components-bcea55bf-ed59-4b11-b80b-6311749da460.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
```
In macOS VoiceOver when the filter mode is `none` and I arrow between items, it reads the item names. But when I set filter mode to `standard`, it doesn't say anything when I arrow between items. I'm guessing it's because focus is still on the search input.

This seems like an important interaction for screenreader users; without I'm not sure how they'd tell what they've selected. (It's possible there are other gestures I'm not aware of) But I tried the same on the combobox and our behavior there is broken too (it reads the previously selected item for some reason).

atmgrifter00 replied:
In my research into getting this to work as expected I found that it is common (if not universal) that the input element where a user enters text applies the `aria-controls` and `aria-activedescendant` attributes with the appropriate element IDs. This alone, however, was not enough to get any screen reader (I used NVDA and Windows Narrator) to provide appropriate voice over as well as manage the expected interactive state.

The extra step I could take that would make the screen readers work well in Chromium-based browsers (but not Firefox) was by setting `type="search"` on the `input` element. The drawback to this, however, was that an 'X' glyph was then displayed by the browser, that did not allow for a browser-agnostic way to remove.

The accessibility tree for what we are doing and what is done in screen reader-friendly selects seems consistent. The main difference that I can see as possibly messing things up is the fact that the elements being navigated between are in the shadow DOM.

Also note that the current FAST `Combobox` has similar issues working with screen readers, so this problem extends beyond our own implementation.

For now I think we should move forward with the current approach and then decide what the appropriate action to take is regarding this problem (more research, filing issues with screen readers, etc...).

jattasNI commented on packages/nimble-components/src/select/tests/select.stories.ts:
```diff
@@ -62,6 +65,12 @@ const optionSets = {
     [ExampleOptionsType.manyOptions]: manyOptions
 } as const;
 
+const filterModeDescription = `
```
We should provide usage guidance about when to set this. I think @fredvisser mentioned in a comment on a previous PR/issue that we'd likely recommend "none" if there are few items and "standard" if there are many items but you'd have to dig that up to find the exact numbers.

atmgrifter00 replied:
Updated docs.

jattasNI reviewed (commented):
Initial review just of storybook behavior and content

mollykreis commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -60,34 +196,203 @@ export class Select extends FoundationSelect implements ErrorPattern {
+                // only filtered option is disabled
+                this.selectedOptions = [];
+                this.selectedIndex = -1;
+                this.filteredOptions[0]!.selected = false;
```
Same question here. Am I missing something? Wouldn't `this.filteredOptions[0]!.selected` already be `false` since you called `this.clearSelection()` above in this function?

mollykreis commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,6 +69,122 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        this.addEventListener('focusout', this.focusoutHander);
+    }
+
+    public override connectedCallback(): void {
+        super.connectedCallback();
+    }
```
It doesn't seem like we need this `connectedCallback` if all it does is call the super class.

mollykreis commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -60,34 +196,203 @@ export class Select extends FoundationSelect implements ErrorPattern {
-        if (this.region && this.control) {
-            this.region.anchorElement = this.control;
+        super.openChanged(prev, next);
+        if (!this.$fastController.isConnected) {
+            return;
+        }
```
All the code that ensures the select is in the correct state is skipped when the element is connected. It seems like there is likely a code path involving disconnecting & reconnecting the element and having `open` change while it is disconnected that could result in the element being in an invalid state.

atmgrifter00 replied:
I have removed this since I don't believe there are any negative consequences to not focusing the input element on a programmatic setting of open to true.

mollykreis commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -102,6 +407,33 @@ export class Select extends FoundationSelect implements ErrorPattern {
     }
+
+    private updateSelectedIndexFromFilteredSet(): void {
+        if (
+            !this.options[this.selectedIndex]
+            && !this.committedSelectedOption
+        ) {
+            return;
+        }
+
+        const selectedItem = this.options[this.selectedIndex] ?? this.committedSelectedOption!;
```
What would you think about rewriting this as:

```suggestion
        const selectedItem = this.options[this.selectedIndex] ?? this.committedSelectedOption;
        
        if (!selectedItem) {
            return;
        }
```

I believe that would keep the same behavior, but it would allow you to get rid of the non-null assertion.

mollykreis commented on packages/nimble-components/src/select/styles.ts:
```diff
@@ -28,6 +40,88 @@ export const styles = css`
+        align-items: center;
+    }
+
+    .no-results-label.below {
+        padding: ${smallPadding} ${mediumPadding} 0 ${mediumPadding};
+    }
+
+    .no-results-label.above {
+        padding: 0 ${mediumPadding} ${smallPadding} ${mediumPadding};
+    }
```
Why aren't these the same styling of `padding: ${smallPadding} ${mediumPadding};`? Perhaps I'm mistaken, but I thought one of the goals was for the height of the drop-down to be the same with the "No items found" text as with a single item. If that's the case, then don't you need the 4px padding on the top & bottom of this span?

rajsite commented on change/@ni-nimble-components-bcea55bf-ed59-4b11-b80b-6311749da460.json:
```diff
@@ -0,0 +1,7 @@
+  "packageName": "@ni/nimble-components",
+  "email": "26874831+atmgrifter00@users.noreply.github.com",
+  "dependentChangeType": "patch"
+}
```
FYI I see a fair bit of open feedback on the PR so I haven't prioritized reviewing it. Wanna make sure we didn't deadlock waiting on each other to review.

atmgrifter00 commented:
@rajsite @jattasNI @mollykreis, I wanted to call out explicitly new behaviors that I have introduced recently as part of this PR. I've already discussed this with both @mollykreis and @jattasNI and we all seem to be in agreement with the behavior. There are two notable new behaviors being introduced (and one minor one):
1. When a user navigates between options in the dropdown using the arrow keys (or by filtering) the display are of the select will _not_ update with the selected value in the dropdown (regardless of the `filterMode` setting). This helps avoid a confusing case for what a user should expect to happen after doing some navigation followed by filtering that removes all options from the dropdown. What should the user see in the display area in this case (particularly when it had been updating earlier as the user was navigating between options). By not updating that display, this confusion is removed.
2. Pressing `Esc` while the dropdown is open when `filter-mode` is set to `none` will _not_ commit the value. I imagine this behavior isn't controversial with the other behavior in place.

Both of these behaviors are actually called out explicitly in the [ARIA guidance for a 'Select-only' style combobx](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-select-only/), as noted departures from the native select input, but done because they improve the accessibility and usability of the component.

A minor new behavior I've introduced that I think is a slight usability improvement is preventing the `Enter` key from dismissing the popup if the user has filtered down to either only disabled options or no options at all. If a user wants to dismiss the popup in this state they can either press `Esc` or click away. Preventing the `Enter` key from dismissing is just a small hint to the user that they actually couldn't commit anything when perhaps that thought they could (perhaps they overtyped and filtered out an option they were meaning to filter to and quickly select).

rajsite commented on change/@ni-nimble-components-bcea55bf-ed59-4b11-b80b-6311749da460.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Adding filter input to Select",
```
Moved to comment thread in-case discussion needed:
@rajsite @jattasNI @mollykreis, I wanted to call out explicitly new behaviors that I have introduced recently as part of this PR. I've already discussed this with both @mollykreis and @jattasNI and we all seem to be in agreement with the behavior. There are two notable new behaviors being introduced (and one minor one):

When a user navigates between options in the dropdown using the arrow keys (or by filtering) the display of the select will not update with the selected value in the dropdown (regardless of the filterMode setting). This helps avoid a confusing case for what a user should expect to happen after doing some navigation followed by filtering that removes all options from the dropdown. What should the user see in the display area in this case (particularly when it had been updating earlier as the user was navigating between options)? By not updating that display, this confusion is removed.
Pressing Esc while the dropdown is open when filter-mode is set to none will not commit the value. I imagine this behavior isn't controversial with the other behavior in place.
Both of these behaviors are actually called out explicitly in the [ARIA guidance for a 'Select-only' style combobx](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-select-only/), as noted departures from the native select input, but done because they improve the accessibility and usability of the component.

A minor new behavior I've introduced that I think is a slight usability improvement is preventing the Enter key from dismissing the popup if the user has filtered down to either only disabled options or no options at all. If a user wants to dismiss the popup in this state they can either press Esc or click away. Preventing the Enter key from dismissing is just a small hint to the user that they actually couldn't commit anything when perhaps that thought they could (perhaps they overtyped and filtered out an option they were meaning to filter to and quickly select).

jattasNI replied:
I think we should consider marking this a `major` change. While it doesn't break any APIs it is a behavior change which could theoretically break tests or corner case user behaviors.

We should also ensure we have Nimble tests for all the behaviors you describe in this comment (I haven't taken a look at the PR updates yet, just a reminder).

jattasNI replied:
One more corner case I noticed during testing (this is pre-existing behavior, not caused by this PR):
1. open select by clicking on it
2. dismiss popup via Enter or Esc
3. select still has keyboard focus so pressing Up/Down causes the value to change. However it doesn't have any visible focus indicator so this was slightly unexpected for me.

I leave it to you to decide what the correct behavior is in this case and whether we should implement that in this PR, file an issue, or treat it as an undefined corner case.

atmgrifter00 replied:
My only non-pushy pushback on the idea that it is a major change, is that the only real functional change (in terms of programmatic observable state changes) is that we will no longer issue a change event when a user presses 'Esc'. If this is enough of a concern or grounds to make it a major change that is fine with me.

jattasNI replied:
I'll admit I'm on the fence and am willing to be outvoted, but my rationale for making it a major change is that it's plausible (though admittedly unlikely) that the behavior change could break a client test. I believe a client whose test was broken would appreciate the heads up and the cost on other clients of making it a major change isn't high.

rajsite replied:
I think I can see it either way as well

> the only real functional change (in terms of programmatic observable state changes) is that we will no longer issue a change event when a user presses 'Esc'.

The same order of operations now has new meaning, open arrow, esc caused a value change and now open, arrow, esc doesn't cause a value change.

But I think ideally a test outside nimble shouldn't need to couple so highly to those operations. A page object should encapsulate a high level "select option x" and if that existed at that level a test wouldn't have broken. i.e. if the page object being added was already in use.

For an intended action we behave the same before and after. If the user changes an option an event is fired and if the user does not change an option and event is not fired. I think in this case I lean slightly towards not breaking and someone who breaks was overly coupled to implementation details (and we didn't give another choice really yet 😅) but I don't feel strongly.

jattasNI replied:
In the interest of resolving this relatively insignificant debate, I'm fine with `minor`.

jattasNI commented on packages/nimble-components/src/select/testing/select.pageobject.ts:
```diff
@@ -0,0 +1,144 @@
+            );
+        }
+
+        const option = this.selectElement.options[index]!;
```
Does this need to consider the case where the desired option is scrolled off screen? If that's complicated to do but detecting that case is easy, I'd be fine with throwing for now until we have a concrete need for it.

atmgrifter00 replied:
Or we could always scroll it into view (though I'm not sure that's even technically necessary). I'm not sure I know of a good reason to prevent a click on an option that is currently scrolled out of view (via this API).

jattasNI replied:
I agree scrolling into view would be better, just wasn't sure if there's an easy API for that. But now that I spent 15 seconds researching it, [scrollIntoView()](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) sounds pretty darn promising!

jattasNI commented on packages/nimble-components/src/select/testing/select.pageobject.ts:
```diff
@@ -0,0 +1,144 @@
+        await waitForUpdatesAsync();
+    }
+
+    public pressEnter(): void {
```
We don't have concrete guidelines for page object method naming, but in the [rich text editor page object](https://github.com/ni/nimble/blob/main/packages/nimble-components/src/rich-text/editor/testing/rich-text-editor.pageobject.ts) we include the work "key" in the keyboard method names so I'd vote to do that here too. `pressEnterKey()`, `pressArrowDownKey()`, etc.

jattasNI reviewed (commented):
Still just a high-level review, will aim to do thorough implementation review tomorrow.

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -1,17 +1,28 @@
-import { attr, html, observable } from '@microsoft/fast-element';
+import {
```
Having steeped in this a bit I found myself needing to pull up FoundationSelect side-by-side to follow the code because we are overriding so much of the behavior.

Are we getting much value out of not forking the foundation implementation? I can't help but feel we have tipped the scale on this control such that if the FoundationSelect did have significant bug fixes or feature additions we probably won't be accelerated by leveraging those changes but get stuck trying to integrate against them.

I think we would still leverage the FormAssociatedSelect base class and the DelegatesARIASelect mixins. Sorry for coming at it with that kind of discussion so late but if we do go with owning the implementation I wouldn't expect it to be a huge amount of work to inline and merge the changes. But if it is that would be an argument against.

atmgrifter00 replied:
At the moment I'm wary of moving forward with that effort, because I think that's an unknown amount of work. We aren't just inheriting from the FAST Select, but also the FAST Listbox, by virtue of the FAST Select, and we _are_ leveraging some of the work each class does for various situations.

That said, I don't disagree with the sentiment and think we should revisit this.

rajsite replied:
Too be clear, I'm not saying to online the full tree, just the contents of FoundationSelect, above FoundationSelect the hierarchy doesn't change. Ie still inherit from FormAssociatedSelect, etc.

That's where my thought it would not be too hard comes from, is there something In missing there?

atmgrifter00 replied:
Ah...I'll take a look that that.

jattasNI replied:
I'll hold off on re-reviewing until this update.

atmgrifter00 replied:
Done.

jattasNI commented on change/@ni-nimble-components-bcea55bf-ed59-4b11-b80b-6311749da460.json:
```diff
@@ -0,0 +1,7 @@
+{
```
If I open the select by tabbing to it with the keyboard and pressing Enter, the filter input doesn't have focus. And I can't see any gesture to give it focus. Is that intentional?

atmgrifter00 replied:
Nope, that's just a goof...I'll fix that.

jattasNI commented on change/@ni-nimble-components-bcea55bf-ed59-4b11-b80b-6311749da460.json:
```diff
@@ -0,0 +1,7 @@
+{
```
Reminder to update the PR description to capture some of the more major changes in the PR.

1. what was copied from FAST and why
2. what changes were made to existing select behavior and why

jattasNI commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -1,31 +1,83 @@
+
 /**
- * A nimble-styled HTML select
+ * A nimble-styled HTML select. The FAST Select implementation has largely been
```
nit: this commentary about FAST shouldn't appear in the class JSDocs as they are client-facing but that info is an implementation detail. You could follow the pattern we've used in other forks of putting a comment at the top of the file or somewhere else that's not public and JSDoc syntax.

jattasNI commented on change/@ni-nimble-components-bcea55bf-ed59-4b11-b80b-6311749da460.json:
```diff
@@ -0,0 +1,7 @@
+{
```
@atmgrifter00 @rajsite I'm finding it difficult to review this PR now since it's adding several hundred lines of FAST code with custom implementations sprinkled in throughout. Should we break it into separate PRs? Annotate what's custom and what's copied? Any other suggestions?

rajsite replied:
That change was a result of the discussion in this thread: https://github.com/ni/nimble/pull/1762#discussion_r1475489622

Summarized as we should own the code instead of overriding tons of Foundation methods. But we should sync up.

rajsite replied:
Talked at standup and meyers is adding some PR comments to guide what changes are adopted from the fast impl

jattasNI replied:
The comments were really helpful, thank you.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -99,14 +165,39 @@ export class Select extends FoundationSelect implements ErrorPattern {
         super.connectedCallback();
         this.addEventListener('change', this.changeValueHandler);
+        this.addEventListener('contentchange', this.updateDisplayValue);
+        this.forcedPosition = !!this.positionAttribute;
```
The above two lines copied from FAST Select.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -33,10 +51,28 @@ declare global {
+     * HTML Attribute: open
+     */
+    @attr({ attribute: 'open', mode: 'boolean' })
+    public open = false;
```
Copied from FAST Select.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,51 +146,689 @@ export class Select extends FoundationSelect implements ErrorPattern {
+     * @internal
+     */
+    @volatile
+    public get collapsible(): boolean {
```
It's possible that we could remove this member as it's intended to make the `Select` look more like a `Listbox`.

jattasNI replied:
I was about to vote to remove it but it looks like it's referenced a lot so probably not worth the effort to do so in this PR.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,51 +146,689 @@ export class Select extends FoundationSelect implements ErrorPattern {
+
+        this.maxHeight = this.position === SelectPosition.above
+            ? Math.trunc(currentBox.top)
+            : Math.trunc(availableBottom);
```
@jattasNI @rajsite, all of the above new implementation in this method. just copied directly from FAST.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,51 +146,689 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        this.updateValue();
+        // We need to force an update to the filteredOptions observable
+        // (by calling 'filterOptions()) so that the template correctly updates.
+        this.filterOptions();
```
@jattasNI @rajsite, everything above this line is copied from FAST Select.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,51 +146,701 @@ export class Select extends FoundationSelect implements ErrorPattern {
+
+            if (!captured?.disabled) {
+                this.updateSelectedIndexFromFilteredSet();
+            }
```
@jattasNI @rajsite, this`if` block is new, everything else above here is copied from FAST.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,51 +146,701 @@ export class Select extends FoundationSelect implements ErrorPattern {
+
+        this.open = this.collapsible && !this.open;
+
+        this.focusFilterInput();
```
@jattasNI , @rajsite this line is new, everything else is from FAST.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,51 +146,701 @@ export class Select extends FoundationSelect implements ErrorPattern {
+     * @param prev - the previous multiple value
+     * @param next - the current multiple value
+     */
+    public override multipleChanged(
```
@jattasNI @rajsite, the three overrides above are copied directly from FAST.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,51 +146,701 @@ export class Select extends FoundationSelect implements ErrorPattern {
+    }
+
+    public override focusoutHandler(e: FocusEvent): BooleanOrVoid {
+        this.updateSelectedIndexFromFilteredSet();
```
@jattasNI @rajsite, everything below this line in this method is copied directly from FAST.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,51 +146,701 @@ export class Select extends FoundationSelect implements ErrorPattern {
+                }
+                break;
+            }
+            case keyTab: {
```
This case was copied directly from FAST.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,51 +146,701 @@ export class Select extends FoundationSelect implements ErrorPattern {
+     *
+     * @internal
+     */
+    public override selectedIndexChanged(
```
@jattasNI @rajsite, everything from here until `openChanged` _except_ for `setSelectedOptions` is copied from FAST.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,51 +146,701 @@ export class Select extends FoundationSelect implements ErrorPattern {
+     * @override
+     * @internal
+     */
+    protected override selectedOptionsChanged(
```
@jattasNI @rajsite, from here until `clearSelection` _except_ for `filterOptions` is copied directly from FAST.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,51 +146,701 @@ export class Select extends FoundationSelect implements ErrorPattern {
+     * @internal
+     * @override
+     */
+    public override handleChange(source: unknown, propertyName: string): void {
```
@jattasNI @rajsite, this and the following 2 overrides are copied directly from FAST.

jattasNI commented on packages/nimble-components/src/select/tests/select.spec.ts:
```diff
@@ -258,4 +304,225 @@ describe('Select', () => {
         });
     });
+
+    describe('filtering', () => {
```
Could you make sure we have tests that verify each of these cases? I've noticed issues with some of them at various iterations of the implementation and if those issues are worth fixing then they're worth testing for. Others are things I noticed in the implementation but didn't see covered.

1. which keys can be used to open the dropdown
1. the focus state of the filter input after opening the dropdown both via the mouse and via the keyboard
1. the focus visible state of the select after various operations which close the drop down (Enter, Esc, click on an option, click outside) with both filter modes
1. the ARIA roles/states of the component and its parts
1. other attributes that are configured in the template like the placeholder text and disabled state of the filter input
1. changing the slotted options while the filter is applied should apply the filter to the new options
1. clicking in the filter input shouldn't close the dropdown
2. space should select an item if filterMode is none

atmgrifter00 replied:
I've added tests to cover _most_ of what you've identified as gaps. The one thing that I still have not found a way to successfully test is the focus state of the filter input after opening the dropdown. For now I suggest we move forward and just concede that there doesn't seem to be a practical way to test this (or if there is it's not worth delaying the feature to discover what the proper means is).

jattasNI reviewed (commented):
Didn't quite finish reviewing the index file but wanted to get my comments submitted in case I don't have a chance to re-review. I reviewed most of it so you should be ok to complete the PR before I return if other reviewers approve.

jattasNI commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -49,51 +146,701 @@ export class Select extends FoundationSelect implements ErrorPattern {
+
+    private focusFilterInput(): void {
+        if (this.open && this.filterMode !== FilterMode.none) {
+            window.requestAnimationFrame(() => {
```
Do we have policies about using `requestAnimationFrame()` vs `DOM.queueUpdate()` or other update queue options? I can't recall but I know we don't use rAF very often so wanted to point it out.

atmgrifter00 replied:
Yes, this has been a pain point with this implementation and has been discussed with @rajsite and @mollykreis independently. Ultimately, there was no other _practical_ known way to successfully get the focus transferred to the filter input (the other way that worked was a `DOM.queueUpdate()` wrapped in an `DOM.queueUpdate()`). We also found a usable within FAST that also did a rAF to accomplish a specific task (their `Listbox` implementation to scroll an element into view`), so it seemed to have precedent.

jattasNI reviewed (commented):
Feel free to submit before I'm back if other reviewers approve.

atmgrifter00 commented on packages/nimble-components/src/select/tests/select.spec.ts:
```diff
@@ -258,4 +304,225 @@ describe('Select', () => {
+            expect(element.value).toBe('one');
+        });
+
+        it('after pressing <Esc> to close dropdown, <Enter> will re-open dropdown', async () => {
```
Fixed (with tests).

atmgrifter00 commented on packages/nimble-components/src/select/tests/select.spec.ts:
```diff
@@ -258,4 +318,339 @@ describe('Select', () => {
+            expect(document.activeElement!.shadowRoot!.activeElement).toBe(
+                element.filterInputElement!
+            );
+        });
```
@rajsite @jattasNI, after chatting with @mollykreis we realized that the proper way to test for focus (as we've discovered before, and I just forgot) is to test the `activeElement` of the `shadowRoot` of the current `activeElement` if the focus was given to something in the shadow DOM.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +89,776 @@ export class Select extends FoundationSelect implements ErrorPattern {
+     */
+    public regionLoadedHandler(): void {
+        this.focusAndScrollOptionIntoView();
+    }
```
@rajsite @jattasNI @mollykreis, I've added this handler is response to noticing that in my attempts to integrate the filterable select into SLE, in the instance I was updating the text focus was not given to the filter input the first time you opened the dropdown (a fairly significant problem). By adding this handler, coupled with a style change I already had in place to use `display: none` instead of `visibility: hidden` to control the visibility of the dropdown, I was able to get the `Select` to behave as expected. Note that in our own Angular example app, I was _not_ able to replicate the same problem.

Unfortunately, I'm not aware of a good way to test this specific implementation, particularly with respect to solving the issue at hand. I propose we move forward with this implementation anyway, as it should be innocuous at worst.

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +89,776 @@ export class Select extends FoundationSelect implements ErrorPattern {
+    private forcedPosition = false;
+    private indexWhenOpened?: number;
+
+    // This intersection observer is to handle focus behavior for when the dropdown becomes
+    // visible.
```
```suggestion
    // This intersection observer is used to wait for anchored region styles to resolve to visible
    // before programmatically calling focus and scrolling selected options into view
```

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +89,776 @@ export class Select extends FoundationSelect implements ErrorPattern {
+
+    public override connectedCallback(): void {
+        super.connectedCallback();
+        this.addEventListener('change', this.changeValueHandler);
+        this.addEventListener('contentchange', this.updateDisplayValue);
```
Can we do normal template binding to these events?

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +89,776 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        this.removeEventListener('change', this.changeValueHandler);
+        this.removeEventListener('contentchange', this.updateDisplayValue);
+        this.regionElementIntersectionObserver.unobserve(this.region);
+        super.disconnectedCallback();
```
Is the ordering significant? If so can youu add a comment, if not can we call super first?

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +89,776 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        this._options = value;
+        Observable.notify(this, 'options');
+    }
+
+    // NOTE: This is a copy of the parent implementation. When providing an override
+    //  for a property setter, you must also provide its corresponding getter.
```
```suggestion
    // This is copied directly from FAST's implemention of its Select component.
```

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +89,776 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        let newValue = next;
+
+        if (this.options?.length) {
+            const selectedIndex = this.options.findIndex(
```
Maybe we can call this `selectedFilteredIndex` to distinguuish from `this.selectedIndex`?

atmgrifter00 replied:
Eh, this isn't necessarily indicative of a filtered index, as this code path is valid for the non-filter case. Perhaps `newValueIndex`?

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +89,776 @@ export class Select extends FoundationSelect implements ErrorPattern {
+    // This is needed because while the dropdown is open the current 'selectedIndex' (set
+    // within this implementation) needs to be relative to the filtered options.
+    public override set value(next: string) {
+        const prev = `${this._value}`;
```
Why is the wrapping template string needed?

```suggestion
        const prev = this._value;
```

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +89,776 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        { threshold: 1.0, root: document }
+    );
+
+    public override connectedCallback(): void {
```
Can you make a pass of all the public properties and methods and make sure they are correctly marked public vs internal?

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +89,776 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        }
+    }
+
+    public setPositioning(): void {
```
does this need to be public?

atmgrifter00 replied:
Nope.

rajsite commented on packages/nimble-components/src/select/styles.ts:
```diff
@@ -28,6 +40,81 @@ export const styles = css`
+        flex-flow: column-reverse;
+    }
+
+    .filter-field {
+        display: flex;
+        flex-direction: row;
+        align-items: center;
+        height: ${controlHeight};
+        background: transparent;
+    }
+
+    .filter-icon {
+        padding-left: ${smallPadding};
+    }
+
+    .filter-field::after,
+    .filter-field::before {
+        content: '';
+        position: absolute;
+        height: 0px;
+        border-bottom: rgba(${borderRgbPartialColor}, 0.1) 2px solid;
+    }
+
+    .filter-field::after {
+        top: calc(${controlHeight} + ${smallPadding} - ${borderWidth});
+    }
+
+    .filter-field:before {
+        bottom: calc(${controlHeight} + ${smallPadding} - ${borderWidth});
+    }
+
+    .filter-field:not(.above)::after,
+    .filter-field.above::before {
+        width: calc(100% - (2 * ${borderWidth}));
+    }
```
Can you group these based on [dom order](https://github.com/ni/nimble/blob/main/packages/nimble-components/docs/css-guidelines.md) and target element instead of similar styles?


![image](https://github.com/ni/nimble/assets/1588923/119db136-8f9a-46e0-a304-50513fa70bcf)


```
.filter-field {}
.filter-field::before {}
.filter-field.above::before {}
.filter-field::after {}
.filter-field:.above::after {}

.filter-icon {}
.filter-input {}
.filter-input::placeholder {}
.filter-input:focus-visible{}
```

atmgrifter00 commented on packages/nimble-components/src/patterns/dropdown/styles.ts:
```diff
@@ -128,14 +128,18 @@ export const styles = css`
         border-color: rgba(${borderRgbPartialColor}, 0.1);
     }
 
+    :host([disabled]) .anchored-region {
+        color: ${bodyDisabledFontColor};
+    }
```
It can be opened if the `open` attribute is set.

rajsite replied:
chatted with meyers and there isn't a clear use-case for a dialog being open and disabled at the same time. A tech debt will be created to address this for all our popover controls.

atmgrifter00 replied:
- #1823

atmgrifter00 commented on packages/nimble-components/src/select/template.ts:
```diff
@@ -81,21 +87,50 @@ SelectOptions
+                    <div ${ref('scrollableElement')}
+                        class="scrollable-element">
+                        <slot
+                            ${slotted({
         filter: (n: Node) => n instanceof HTMLElement && Listbox.slottedOptionFilter(n),
         flatten: true,
         property: 'slottedOptions',
     })}
```
I can disable the rule for the file. Not sure if there are other viable options.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +89,776 @@ export class Select extends FoundationSelect implements ErrorPattern {
+    @observable
+    public filter = '';
+
+    /**
+     * @internal
+     */
+    @observable
+    public scrollbarIsVisible = false;
```
Nope. This was an implementation vestige of what was needed to support the ideal visual design that Brandon created.

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +89,776 @@ export class Select extends FoundationSelect implements ErrorPattern {
+                entries.length > 0
+                    && entries[entries.length - 1]!.intersectionRatio > 0
+            ) {
+                this.focusAndScrollOptionIntoView();
```
No, if it is collapsed, this should be a no-op. If the popup is open this just means that if the selected option had been scrolled out of view, it will be scrolled into view, which seems like an okay behavior.

atmgrifter00 commented on packages/nimble-components/src/select/template.ts:
```diff
@@ -31,8 +37,8 @@ SelectOptions
-        aria-activedescendant="${x => x.ariaActiveDescendant}"
-        aria-controls="${x => x.ariaControls}"
+        aria-activedescendant="${x => (x.filterMode === FilterMode.none ? x.ariaActiveDescendant : null)}"
+        aria-controls="${x => (x.filterMode === FilterMode.none ? x.ariaControls : null)}"
```
Agreed. I will follow up with this in a separate PR. Have created the following issues:
- #1821 
- #1822

rajsite commented on packages/nimble-components/src/select/tests/select.spec.ts:
```diff
@@ -258,4 +318,339 @@ describe('Select', () => {
+                name: 'standard'
+            }
+        ];
+        filterModeTestData.forEach(testData => {
```
And can you organize the test as parameterized suites?

```ts
filterModeTestData.forEach(testData => {
    describe(`with filterMode = ${testData.name}`, () => {
        it(... /* now each test does not need to say it's filter mode since it is part of the describe block */)
        it(...)
    })

})
```

atmgrifter00 replied:
Created #1827.

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +88,801 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        { threshold: 1.0, root: document }
+    );
+
+    public override connectedCallback(): void {
+        super.connectedCallback();
+        this.forcedPosition = !!this.positionAttribute;
+        this.anchoredRegionIntersectionObserver.observe(this.anchoredRegion);
+        this.initializeOpenState();
+    }
+
+    public override disconnectedCallback(): void {
+        super.disconnectedCallback();
+        this.anchoredRegionIntersectionObserver.unobserve(this.anchoredRegion);
+    }
```
these should be internal

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +88,801 @@ export class Select extends FoundationSelect implements ErrorPattern {
     }
 
-    // Workaround for https://github.com/microsoft/fast/issues/5773
+    public get displayValue(): string {
```
should this be internal?

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +88,801 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        return true;
+    }
+
+    public override focusoutHandler(e: FocusEvent): BooleanOrVoid {
```
should be internal

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +88,801 @@ export class Select extends FoundationSelect implements ErrorPattern {
+    }
+
+    /**
+     * @ainternal
```
typo

rajsite commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +88,801 @@ export class Select extends FoundationSelect implements ErrorPattern {
+    }
+
+    /**
+     * @ainternal
```
typo

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +88,801 @@ export class Select extends FoundationSelect implements ErrorPattern {
+     * @internal
+     */
+    public override disabledChanged(prev: boolean, next: boolean): void {
+        if (super.disabledChanged) {
```
This is because this is part of the [`FormAssociated` interface](https://github.com/microsoft/fast/blob/c64110d2f1dfbc24a0978c362bcf72b8958e6d73/packages/web-components/fast-foundation/src/form-associated/form-associated.ts#L121) which types this to a nullable callback.

rajsite replied:
Excellent, thanks for checking 🙏

rajsite commented on change/@ni-nimble-components-bcea55bf-ed59-4b11-b80b-6311749da460.json:
```diff
@@ -0,0 +1,7 @@
+{
```
Oh it's subtle, here is a slowed down gif of it
![flickermenu](https://github.com/ni/nimble/assets/1588923/492f8e84-995f-4cfa-8e0b-bb440b3cab0c)

atmgrifter00 replied:
Yep. So this is a result of using `display: none` instead of `visibility: hidden` to control the visibility of the dropdown. I've now reverted that change, removed the `IntersectionObserver`, and put back in the `requestAnimationFrame`. All of this along with the existing `loaded` event handler for the `anchoredRegion` still seems enough to make the integration into SLE use-case continue to work.

rajsite replied:
If the `    .anchored-region[hidden] { visibility: hidden;}` can we still move it to the anchored region styles?
We should also have a comment that says we are intentionally not using the display helper.

```
    :host {
        ${'' /* Avoid using the `display` helper to customize hidden behavior }
        display: block;
        contain: layout;
        z-index: ${ZIndexLevels.zIndex1000};
    }
    :host[hidden] {
        visibility: hidden;
    }
```

atmgrifter00 commented on packages/nimble-components/src/select/index.ts:
```diff
@@ -39,61 +88,801 @@ export class Select extends FoundationSelect implements ErrorPattern {
+        Observable.notify(this, 'options');
+    }
+
+    // This is copied directly from FAST's implemention of its Select component.
```
Removed.

atmgrifter00 commented on change/@ni-nimble-components-bcea55bf-ed59-4b11-b80b-6311749da460.json:
```diff
@@ -0,0 +1,7 @@
+{
```
Yeah, I think we can revisit this in a follow-up PR if we feel this is wonky.
